### PR TITLE
Upload images in admin with drag and drop

### PR DESCRIPTION
**Pull request by mtomov**
I wanted to see how much effort it would be to do a modern drag and drop image upload (#1518), and that pretty much does it. Code is ~~not~~ ideal, but wanted to post it here if somebody wants to try it out, have any comments, etc ..

In summary:
- The JavaScript is just about ~~100 lines~~ (grew to ~ 200 with Backbone), even non-coffee-script (as I started with a non coffee example, but can convert it around). 
- Code I started from includes tests for feature detection (Drag and Drop, File Upload, Progress Bar). I decided to leave them for reference, but they can be removed, as I think that admin users should run the very latest of browsers, and if they don't - they can use the old functionality.
- All of the old way of uploading images is in place, so if someone is stuck with some unsupported functions, they can always fallback to that.
- There's a `javascript` response from the server, which appends a row in the images table .. not sure if it's an accepted practice
- I need to add a capybara test, but shouldn't be difficult, as I have also left in place a Browse button if somebody doesn't want to drag and drop
- Image is automatically assigned the master variant
  - so to make it a truly nice experience, the rows in the images table will have an select dropdown field with ability to change the variant, and click a Tick button on the right, similarly to the Stock Items page, and other in-place row edits.

And some pictures:

screenshot from 2016-10-27 23 30 25

screenshot from 2016-10-27 23 40 48

or in boxes:

screenshot from 2016-11-03 21 40 05

### CODE REVIEW

**Comment by mtomov** [comment]
Hello,

This one is ready for official review. I've cleaned up the javascript using Backbone, and have placed the supporting bits (css, js, translations) to their relevant places.

This is first time I'm using Backbone and I'm quite pleasantly surprised how it put a nice structure overall. The events mapping is very clean, and the view and its model work very well together to provide for _live_ data binding.

You might've noticed that I'm using the _underscore's_ templating function `_template()` instead of the customary Handlebars, and the reason is that I don't like how all handlebar templates are placed in a god forsaken directory separated from their actual location of usage. In terms of performance - since the template is only parsed and evaluated once on the initial page load, and I'm using the simpler template syntax with Underscore, I don't think there is a negative burden on the page.

In terms of design - current one is a bit plain. An alternative is to make use of [Bootstrap's Cards](https://v4-alpha.getbootstrap.com/components/card/#groups). So, either multiple images per row, or again a single image per row, but [wrapped in a round border](https://v4-alpha.getbootstrap.com/components/card/#outline-variants).

Thank you!

**Comment by tvdeyen** [comment]
@mtomov thanks! I will have a look later.

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+      // send the image to the server
+      Spree.ajax({
+        url: window.location.pathname,
+        type: "POST",
+        dataType: 'script',
+        data: formData,
+        processData: false,  // tell jQuery not to process the data
+        contentType: false,  // tell jQuery not to set contentType
+        xhr: function () {
```
Comment (Sinetheta) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: Adding progress to ajax is cool, we may want to make this generic to all `Spree.ajax` in the future. Make it a callback and trigger an event.

File: backend/app/views/spree/admin/images/_image_row.html.erb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<tr id="<%= spree_dom_id image %>" data-hook="images_row" class="<%= cycle('odd', 'even')%>">
```
Comment (Sinetheta) [inline backend/app/views/spree/admin/images/_image_row.html.erb]: As we add more long lived pages `cycle('odd', 'even')` is really gonna show it's limitations, as new uploads don't get striped.

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+      var progressModel = new ProgressModel({file: file});
+      progressModel.previewFile();
+      progressModel.uploadFile();
+
+      var progressView = new ProgressView({model: progressModel});
+      this.progressZone.appendChild(progressView.render().el);
+    },
+
+    onDragOver: function() {
+      this.el.className = 'hover';
```
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: Right so. I've gone for that `dragClass: 'with-images'`

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      return false;
```
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: Good observation. Corrected to that

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    },
+
+    onDragLeave: function() {
+      this.el.className = '';
```
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: Indeed so. Used the native's `el.classList.remove/add`

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+      return false;
```
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: Thanks for looking in those details. Corrected now.

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    },
+
+    onDrop: function(e) {
+      this.el.className = '';
+      e.preventDefault();
+
+      for (var i = 0; i < e.originalEvent.dataTransfer.files.length; i++) {
```
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: did it as per your example - all excellent working. Thanks!

File: backend/app/assets/javascripts/spree/backend/images/upload.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      }
+    },
+
+    onFileBrowserSelect: function(e) {
+      for (var i = 0; i < e.target.files.length; i++) {
+        this.upload(e.target.files[i]);
+      }
+    },
+
+    tests: {
```
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: I'm not in favour of adding more libs. The code for those random pieces of text is ~20 links. As you see fit..

File: backend/app/assets/javascripts/spree/backend/images/upload.js (hunk 1 above)
Comment (mtomov) [inline backend/app/assets/javascripts/spree/backend/images/upload.js]: I think it's too early at this point for anything such. As far as I understand, one only gets progress for large user submitted requests, i.e. files. For regular post requests when the server is just slow .. that's a different story that would like need server support.

**Comment by mtomov** [comment]
Thanks @Sinetheta and John for the elaborate look. I've indeed corrected all points that you mentioned - switched to Handlebars, etc ..

Furthermore:

1. In 8887673 - I added a small piece out of stack overflow, which allows to resolve keys with dots inside. Ex `{{t 'admin.images.index.image_process_failed' }}` - that wasn't possible before

2. In 61a62ec - I've given it a try to use Bootstrap's cards with image inside. I think it looks better. From code perspective only thing I don't like is that I had to place `className: 'col-sm-6 col-md-4 margin-bottom-one',` inside the javascript due to Backbone wrapping the element in a div. 

  I've uploaded image at the top of how the page looks with 3 boxes per row. Naturally, I can get rid of the commit if it's too much. Thanks again!

**Comment by Sanjana6221** [comment]
As I can see when I try to upload large image around 20 MB I see the progress bar show 100% but the loader still running, and when I check it with the server logs it looks like paperclip is still trying to create that file.

Could you please let me know is this working as expected as it implemented?

**Pull request merged**